fix: resetting offline players

leumasme opened the pull request:
Reset offline players the next time they join by tracking a global incrementing current-reset-id in the config.  
Player data (via a playerbound PersistentDataContainer) tracks the `last_reset_id`, that being the ID of the last reset that the player was present for.  
On join, if the current-reset-id > the players last reset id, reset the player.  
This is more elegant than just deleting the playerdata folder as suggested in the other issue, as this allows using the same reset procedure for online and offline players.

Tested to work across server restarts (player leave -> world reset -> server restart -> player join)

[EDIT: see comment] This has an issue with players on mounts/vehicles. An offline player riding a boat will stay in the boat and keep their xyz-position from the old world when getting reset.   
`player.isInsideVehicle()` also seems to return false in `PlayerJoinEvent`, maybe the player(s mount) isn't properly loaded here yet? Maybe resetting on join would need to be delayed by a tick?  
This doesnt affect the existing good behavior of online players, so maybe acceptable for now.

Fixes #6
Fixes #2
Closes https://github.com/qixils/Fahare/pull/5 which was another PR for this that was abandoned and did not persist on server restart

leumasme committed:
fix: resetting offline players

Fixes #6
Fixes #2

leumasme committed:
fix: offline players shouldn't keep mount on reset

leumasme commented:
issue with keeping mount+position is resolved by storing the resetting players uuid on join for 1 tick and then removing the mount entity in the EntityMountEvent that is fired before the tick is over.  
this causes a log message from vanilla: `Couldn't reattach entity to player`  
but i couldn't find a nicer way to do this (without mixins).  
the mounting happens in `ServerPlayer.loadAndSpawnParentVehicle` called from `PrepareSpawnTask.Ready.spawn`.  
The mount is stored in the player NBT under `.RootVehicle.Entity`, but I dont see any Events that receive the player nbt before it is used.

qixils commented:
Thanks for your contributions! I'll get a couple things merged in the morning ❤️

qixils committed:
feat: gradleify & bump deps

ide refused to work with groovy for whatever reason

qixils committed:
feat: read server.properties

qixils committed:
Merge branch 'main' into fork/leumasme/reset-offline-players

qixils committed:
fix: compile against paper hard fork

qixils committed:
fix: avoid saving reset id in config

this will wipe comments from the config; prefer to store in a separate file. lazy and used a txt but can expand into a json later

qixils commented:
Thanks for the contribution!
